local/rest_api_gcbm/gcbm.py

```diff
@@ -0,0 +1,566 @@
+                self.json_paths.update(self.classifier._update_config())
+                deduced_category = "classifier"
+            elif self.miscellaneous._append(filename):
+                self.json_paths.update(self.miscellaneous._update_config())
+                deduced_category = "miscellaneous"
+            elif self.db._append(filename):
+                self.db._update_config()
+                deduced_category = "db"
+            else:
+                raise ValueError(
+                    f"Couldn't deduce a valid from the file path {file_path}, supported categories: {self.supported_categories}"
+                )
+            print(
+                    f"Category wasn't provided, deduced category: {deduced_category} it from the file path: {file_path}"
+            )
+
+    def generate_configs(self):
+        self._generate_provider_config(self.dirpath)
+        self._generate_modules_cbm_config(self.dirpath)
+        self._add_variables(self.dirpath)
+
+    def sync_config(self, file_path):
+        def _write_to_file(file_path, data):
+            with open(file_path, "w+") as _file:
+                _file.write(data)
+
+        data = GCBMSimulation.safe_read_json(file_path)
+
+        if self.files[file_path] != data:
+            # Means data has changed, so update the file_path
+            _write_to_file(file_path, data)
+            # Also update the dict
+            self.files[file_path] = data
+
+    # TODO (@ankitaS11): We can just have these as class methods later, this will reduce the redundancy in the code later
+    def update_disturbance_config(self):
+        self.json_paths.update(self.disturbance._update_config())
+
+    def update_classifier_config(self):
+        self.json_paths.update(self.classifier._update_config())
+
+    def update_miscellaneous_config(self):
+        self.json_paths.update(self.miscellaneous._update_config())
+
+    # TODO: category should be an enum instead of a string to avoid any mistakes
+    def set_attributes(self, category: str, file_path: str, payload: dict):
+        base_path = os.path.basename(file_path)
+        if category == "disturbance":
+            self.disturbance.setattr(
+                file=base_path, attributes=payload, category=category
+            )
+        elif category == "classifier":
+            self.classifier.setattr(
+                file=base_path, attributes=payload, category=category
+            )
+        elif category == "miscellaneous":
+            self.miscellaneous.setattr(
+                file=base_path, attributes=payload, category=category
+            )
+        else:
+            raise ValueError(
+                f"Expected a valid category name out of {self.supported_categories}, but got {category}. Note that setting attributes for a database file is not supported."
+            )
+
+    @staticmethod
+    def safe_read_json(file_path):
+
+        # TODO: add read method for gcbm_config.cfg and logging.conf
```

Comment:
@aornugent - Thanks for adding these TODOs in the code, they are extremely useful for me to track what needs to be done. However, I'm a bit confused about this one - do we want to just read these files, and store them in another folder? Or in a JSON? Or do we want to extract any information, and _use_ it somewhere else? I'm not sure as I couldn't find a relevant part of it in `app.py`, so I'll really appreciate it if you can help me with this.

Reply:
Hi @ankitaS11 - this one is a little cryptic. The two files need to be copied from `templates` and the default behaviour was to traverse all the config. files in the folder. 

But `safe_read_json()` only reads JSON, and was erroring out when given the `.cfg` and `.conf` files. 

The contents of these files don't change very often (if ever!) but presently we have no way of reading them as part of the simulation configuration. 

I'd appreciate your advice on how they can be handled.

Reply:
Hi @aornugent.  I'm not Ankita, but I think [configparser](https://docs.python.org/3/library/configparser.html) library would be a good choice for parsing the `.conf` files. They provide easy parsing for `.ini` style syntax.

For `.cfg` files, we can prepend a pseduo-section and then use `configparser`.

The code would look something like this:
```python
import configparser

# For conf files
parser = configparser.ConfigParser()
parser.read('what.conf')
log = parser['Core']['DisableLogging'])

# For cfg files
parser = ConfigParser()
with open("gcbm_config.cfg") as config:
    parser.read_string("[top]\n" + config.read())
```

The problem is that all keys in the `gcbm_config.cfg` have the same name. This causes an error too.
```
config=localdomain.json
config=pools_cbm.json
config=modules_cbm.json
config=modules_output.json
config=spinup.json
config=variables.json
config=internal_variables.json
```

we can maybe rename some fields? Like `localdomain_config=localdomain.json`? :)

Reply:
Thanks @Crystalsage - that's a good suggestion. I don't think we can change the field names, because they're in the format that the GCBM understands.

Maybe we can just read these files in as strings for now?

Reply:
@aornugent Ah I see. In that case - what do you think about preparing a makeshift `.cfg` config in memory for better parsing and using that throughout the API? This way, we won't need to restructure the config file - and we can always change a few things around the API.  A very rough sketch looks something like this:

```python
import configparser

parser = configparser.ConfigParser()

gcbm_config_file = open("gcbm_config.cfg").read()
config_file_new = ""

# Build a makeshift config
for config in config_file.splitlines():
    # Get the json file name
    file_name = config.split('=')[1]

    # Trim off the extension to get the config type
    config_name = file_name[:-5]

    # Format into .ini style
    # e.g. localdomain_config=localdomain.json
    config_file_new += f"{config_name}_config={file_name}\n"

parser.read_string("[top]\n" + config_record)

# Returns 'localdomain.json'
parser['top']['localdomain_config']
```

Reply:
Hi @Crystalsage  - sorry for the delay. Your proposal looks great!

Reply:
Hi @Crystalsage - are you append your suggested changes to this branch please?